ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java

```diff
@@ -80,30 +104,91 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
+      NetworkRequest unqueuedRequest;
+      while ((unqueuedRequest = unqueuedRequests.poll()) != null) {
+        networkRequestBundle.getRequestsToDrop().add(unqueuedRequest);
+      }
+      requestToServe = networkRequestBundle.getRequestToServe();
+      requestsToDrop = networkRequestBundle.getRequestsToDrop();
+
+      // Consume the first 8 bytes of requests since TCP impl uses this size to allocate buffer.
+      if (requestToServe != null) {
+        http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - requestToServe.getStartTimeInMs());
+        if (!consumeStream(requestToServe)) {
+          requestToServe = null;
+        }
+      }
+      Set<NetworkRequest> errorRequests = new HashSet<>();
+      for (NetworkRequest requestToDrop : requestsToDrop) {
+        http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - requestToDrop.getStartTimeInMs());
+        if (!consumeStream(requestToDrop)) {
+          errorRequests.add(requestToDrop);
+          logger.error("Error reading the stream, dropping the request: {}", requestToDrop);
+        }
+      }
+      requestsToDrop.removeAll(errorRequests);
+    } while (requestToServe == null && requestsToDrop.isEmpty());
+
+    return new NetworkRequestBundle(requestToServe, requestsToDrop);
+  }
+
   @Override
-  public NetworkRequest receiveRequest() throws InterruptedException {
-
-    NetworkRequest request = null;
-    while (request == null) {
-      request = requestQueue.take();
-      http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - request.getStartTimeInMs());
-      if (request.equals(EmptyRequest.getInstance())) {
-        logger.debug("Request handler {} received shut down command ", request);
-      } else {
-        DataInputStream stream = new DataInputStream(request.getInputStream());
-        try {
-          // The first 8 bytes is size of the request. TCP implementation uses this size to allocate buffer. See {@link BoundedReceive}
-          // Here we just need to consume it.
-          stream.readLong();
-        } catch (IOException e) {
-          logger.error("Encountered an error while reading length out, close the connection", e);
-          closeConnection(request);
-          request = null;
+  public List<NetworkRequest> getUnqueuedRequests() throws InterruptedException {
+    List<NetworkRequest> requestsToDrop = new ArrayList<>();
+    NetworkRequest unqueuedRequest;
+    do {
+      while ((unqueuedRequest = unqueuedRequests.poll()) != null) {
+        requestsToDrop.add(unqueuedRequest);
+      }
+      if (requestsToDrop.isEmpty()) {
+        // If there are no un-queued requests, wait on the queue.
+        unqueuedRequest = unqueuedRequests.take();
+        requestsToDrop.add(unqueuedRequest);
+      }
+      Set<NetworkRequest> errorRequests = new HashSet<>();
+      for (NetworkRequest requestToDrop : requestsToDrop) {
+        if (!consumeStream(requestToDrop)) {
```

Comment:
1/ why do we consumeStream if we have decided to drop ?
2/ and what do we do with the memory occupied by a request whose stream we cannot consume ? that doesn't seem to get added to the requeststoDrop queue ?

Reply:
can we explicitly release the mem for those erroneous requests ?

Reply:
> 1/ why do we consumeStream if we have decided to drop ?

When we say drop, we actually mean to reject the reject without handling it. Due to that, we still need to consume the first 8 bytes of the stream. Added more comments in javadoc comments on why we are consuming the first 8 bytes.

> can we explicitly release the mem for those erroneous requests ?

Thank you for pointing it out. Released the memory for erroneous requests